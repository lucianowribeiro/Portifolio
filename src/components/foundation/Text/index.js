/* eslint-disable react/jsx-props-no-spreading */
import React from 'react';
import PropTypes from 'prop-types';
import styled, { css } from 'styled-components';
import breakpointsMedia from '../../../theme/utils/breakpointsMedia';
import Link from '../Link';

export const TextStyleMap = {
  titleXS: css`
    font-size: ${({ theme }) => theme.typographyVariants.titleXS.fontSize};
    font-weight: ${({ theme }) => theme.typographyVariants.titleXS.fontWeight};
    line-height: ${({ theme }) => theme.typographyVariants.titleXS.lineHeight};
  `,
  titleMD: css`
    font-size: ${({ theme }) => theme.typographyVariants.titleMD.fontSize};
    font-weight: ${({ theme }) => theme.typographyVariants.titleMD.fontWeight};
    line-height: ${({ theme }) => theme.typographyVariants.titleMD.lineHeight};
  `,
  subTitleXS: css`
    font-size: ${({ theme }) => theme.typographyVariants.subTitleXS.fontSize};
    font-weight: ${({ theme }) => theme.typographyVariants.subTitleXS.fontWeight};
    line-height: ${({ theme }) => theme.typographyVariants.subTitleXS.lineHeight};
  `,
  subTitleMD: css`
    font-size: ${({ theme }) => theme.typographyVariants.subTitleMD.fontSize};
    font-weight: ${({ theme }) => theme.typographyVariants.subTitleMD.fontWeight};
    line-height: ${({ theme }) => theme.typographyVariants.subTitleMD.lineHeight};
  `,
  menuXS: css`
    font-size: ${({ theme }) => theme.typographyVariants.menuXS.fontSize};
    font-weight: ${({ theme }) => theme.typographyVariants.menuXS.fontWeight};
    line-height: ${({ theme }) => theme.typographyVariants.menuXS.lineHeight};
  `,
  menuMD: css`
    font-size: ${({ theme }) => theme.typographyVariants.menuMD.fontSize};
    font-weight: ${({ theme }) => theme.typographyVariants.menuMD.fontWeight};
    line-height: ${({ theme }) => theme.typographyVariants.menuMD.lineHeight};
  `,
  paragraphXS: css`
    font-size: ${({ theme }) => theme.typographyVariants.paragraphXS.fontSize};
    font-weight: ${({ theme }) => theme.typographyVariants.paragraphXS.fontWeight};
    line-height: ${({ theme }) => theme.typographyVariants.paragraphXS.lineHeight};
  `,
  paragraphMD: css`
    font-size: ${({ theme }) => theme.typographyVariants.paragraphMD.fontSize};
    font-weight: ${({ theme }) => theme.typographyVariants.paragraphMD.fontWeight};
    line-height: ${({ theme }) => theme.typographyVariants.paragraphMD.lineHeight};
  `,
};
const TextBase = styled.span`
  ${breakpointsMedia({
    xs: (props) => TextStyleMap[props.mobile],
    md: (props) => TextStyleMap[props.desktop],
  })}
`;

export default function Text({
  tag, mobile, desktop, children, href, ...props
}) {
  if (href) {
    return (
      <TextBase as={Link} href={href} mobile={mobile} desktop={desktop} {...props}>
        {children}
      </TextBase>
    );
  }
  return (
    <TextBase as={tag} mobile={mobile} desktop={desktop} {...props}>
      {children}
    </TextBase>
  );
}

Text.propTypes = {
  tag: PropTypes.string,
  mobile: PropTypes.string,
  desktop: PropTypes.string,
  href: PropTypes.string,
  children: PropTypes.node,
};

Text.defaultProps = {
  tag: 'span',
  mobile: 'paragraphXS',
  desktop: 'paragraphMD',
  href: '',
  children: undefined,
};
