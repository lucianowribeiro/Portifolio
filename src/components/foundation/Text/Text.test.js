import React from 'react';
import { create } from 'react-test-renderer';
import { render, screen, AllTheProviders } from '../../../infra/test/testUtils';
import Text from './index';

describe('</Text>', () => {
  describe('its render', () => {
    describe('as default', () => {
      test('with props mobile and desktop', () => {
        const childrenMock = 'Teste';
        const component = create(
          <AllTheProviders>
            <Text tag="h1" mobile="titleXS" desktop="titleMD">
              {childrenMock}
            </Text>
          </AllTheProviders>,
        ).toJSON();
        // Assertions
        expect(...component.children).toBe(childrenMock);
        expect(component.type).toBe('h1');
        // Snapshot
        expect(component).toMatchSnapshot();
      });
      test('without props and children', () => {
        // Assertions
        expect(
          create(
            <AllTheProviders>
              <Text />
            </AllTheProviders>,
          ),
        ).toMatchSnapshot();
        //
      });
    });
    describe('as href(Link)', () => {
      test('and props mobile and desktop', () => {
        const childrenMock = 'Teste';
        render(
          <Text href="/" mobile="titleXS" desktop="titleMD">
            Teste
          </Text>,
        );
        // Assertions
        const element = screen.getByRole('link', { name: /teste/i });
        expect(element).toContainHTML(childrenMock);
        expect(element).toHaveAttribute('href', '/');
        //
        expect(element).toMatchSnapshot();
      });
    });
  });
});
