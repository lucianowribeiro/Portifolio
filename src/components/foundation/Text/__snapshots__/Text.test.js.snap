// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`</Text> its render as default with props mobile and desktop 1`] = `
@media only screen and (min-width:0px) {
  .c0 {
    font-size: 7vw;
    font-weight: normal;
    line-height: 1.25;
  }
}

@media only screen and (min-width:652px) {
  .c0 {
    font-size: 3rem;
    font-weight: 500;
    line-height: 1.25;
  }
}

<h1
  className="c0"
>
  Teste
</h1>
`;

exports[`</Text> its render as default without props and children 1`] = `
@media only screen and (min-width:0px) {
  .c0 {
    font-size: 1rem;
    font-weight: normal;
    line-height: 1.25;
  }
}

@media only screen and (min-width:652px) {
  .c0 {
    font-size: 1.2rem;
    font-weight: 400;
    line-height: 1.25;
  }
}

<span
  className="c0"
/>
`;

exports[`</Text> its render as href(Link) and props mobile and desktop 1`] = `
.c0 {
  color: #35474e;
  text-underline-position: under;
}

.c0:hover {
  -webkit-text-decoration: none;
  text-decoration: none;
  color: #009688;
  -webkit-transition: 350ms ease-in-out;
  transition: 350ms ease-in-out;
}

@media only screen and (min-width:0px) {
  .c1 {
    font-size: 7vw;
    font-weight: normal;
    line-height: 1.25;
  }
}

@media only screen and (min-width:652px) {
  .c1 {
    font-size: 3rem;
    font-weight: 500;
    line-height: 1.25;
  }
}

<a
  class="c0 c1"
  href="/"
>
  Teste
</a>
`;
